[quant][pt2e] Add reference representation for quantize_per_channel and dequantize_per_channel

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105784
* __->__ #105783

Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_quantize_dequantize_per_channel

Although right now it is not really testing things since there is some problem with dynamo export

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[quant][pt2e] Add reference representation for quantize_per_channel and dequantize_per_channel

Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_quantize_dequantize_per_channel

Although right now it is not really testing things since there is some problem with dynamo export

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add reference representation for quantize_per_channel and dequantize_per_channel"


Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_quantize_dequantize_per_channel

Although right now it is not really testing things since there is some problem with dynamo export

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add reference representation for quantize_per_channel and dequantize_per_channel"


Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_quantize_dequantize_per_channel

Although right now it is not really testing things since there is some problem with dynamo export

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented on torch/ao/quantization/pt2e/representation/rewrite.py:
```diff
@@ -207,30 +208,81 @@ def _reference_dequantize_per_tensor_int8(x_i8, scale, zero_point, quant_min, qu
+    return out_fp32
+
+def _replace_ph_qdq_per_channel_replacement(gm: torch.fx.GraphModule):
+    return replace_literals_with_existing_placeholders(
```
What is this one doing?

jerryzh168 replied:
will add some comment

jerryzh168 commented on torch/ao/quantization/pt2e/representation/rewrite.py:
```diff
@@ -205,30 +206,81 @@ def _reference_dequantize_per_tensor_int8(x_i8, scale, zero_point, quant_min, qu
-    (_QUANTIZE_PER_TENSOR_INT8_EXAMPLE_INPUTS, _quantize_per_tensor_int8, _reference_quantize_per_tensor_int8, _DONT_REPLACE_LITERAL),
-    (_DEQUANTIZE_PER_TENSOR_INT8_EXAMPLE_INPUTS, _dequantize_per_tensor_int8, _reference_dequantize_per_tensor_int8, _REPLACE_LITERAL),
+    (_QUANTIZED_ADD_EXAMPLE_INPUTS, _qdq_quantized_add_relu, _reference_quantized_add_relu, None, None),
+    (_QUANTIZED_ADD_EXAMPLE_INPUTS, _qdq_quantized_add, _reference_quantized_add, None, None),
```
this is creating a new placeholder automatically, every new literal encountered will be assigned a new placeholder

the other one is using existing placeholder, so we'll need a map from literal to placeholder

jerryzh168 commented on torch/ao/quantization/pt2e/representation/rewrite.py:
```diff
@@ -205,30 +206,81 @@ def _reference_dequantize_per_tensor_int8(x_i8, scale, zero_point, quant_min, qu
+    torch.randn(3, dtype=torch.float),
+    torch.zeros(3, dtype=torch.int),
+    1,
+    -128,
```
why? we only need one example input, but the pattern can be used for other dtypes

andrewor14 commented on torch/ao/quantization/pt2e/representation/rewrite.py:
```diff
@@ -207,30 +208,81 @@ def _reference_dequantize_per_tensor_int8(x_i8, scale, zero_point, quant_min, qu
+    return replace_literals_with_existing_placeholders(
+        gm,
+        exclude_literals=[-1],
+        literal_to_ph_idx={1: 3, -128: 4, 127:5}
```
This seems really hacky. I feel it's actually clearer to just do the manual arg copying if we have to do something like this. How `replace_literals_with_placeholders` works is already a bit complicated to begin with

jerryzh168 replied:
I feel we do not have other solutions before we have a proper fix, and this should be easier than manual arg copying since we do not need to write additional code, how is manual arg copying easier?

kimishpatel replied:
I do feel the same as Andrew, but @andrewor14 can you write an alternate solution? Then we can get an idea whether it makes this any easier. Yes it does look hacky, but also without having alternate solution, it makes it a little harder to compare.

andrewor14 commented on torch/ao/quantization/pt2e/utils.py:
```diff
@@ -261,13 +264,41 @@ def _is_literal(arg):
     return gm
+
+
+def replace_literals_with_existing_placeholders(gm, exclude_literals, literal_to_ph_idx):
```
I don't follow why we need this in addition to the original `replace_literals_with_placeholders`. Can you explain in a comment how this is different from `replace_literals_with_new_placeholders` and the motivation for this? Also what do the args do?

jerryzh168 replied:
sure, added some docs

andrewor14 commented on torch/ao/quantization/pt2e/representation/rewrite.py:
```diff
@@ -207,30 +208,81 @@ def _reference_dequantize_per_tensor_int8(x_i8, scale, zero_point, quant_min, qu
+        literal_to_ph_idx={1: 3, -128: 4, 127:5}
+    )
 
+# (example inputs, pattern, replacement, post_transformation_pattern, post_transformation_replacement)
```
This is getting complicated. What "transformation" are we talking about here? Maybe it makes sense to turn this into a dataclass instead of making the tuples longer

kimishpatel replied:
Yes, in general I feel the whole rewrite feels like not clean, maybe because of subgraph rewriter limitations.

jerryzh168 replied:
done, transformation means some further function to transform the graph, in this case replacing literals with placeholders

andrewor14 reviewed (commented):
Left some comments on the graph replacement part. Overall I feel more inline comments will be helpful, since the new replace literals logic is getting complicated. I'm also wondering if we can just do the manual arg copying to reduce the complexity / make the code easier to maintain.

jerryzh168 committed:
Update on "[quant][pt2e] Add reference representation for quantize_per_channel and dequantize_per_channel"


Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_quantize_dequantize_per_channel

Although right now it is not really testing things since there is some problem with dynamo export

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/pt2e/representation/rewrite.py:
```diff
@@ -205,30 +206,81 @@ def _reference_dequantize_per_tensor_int8(x_i8, scale, zero_point, quant_min, qu
+    return out_i32.to(torch.int8)
+
+_DEQUANTIZE_PER_CHANNEL_INT8_EXAMPLE_INPUTS = (
+    torch.randint(-128, 127, (1, 3, 3, 3), dtype=torch.int8),
```
this is just for generating patterns, I can add some test for 4/5 dims

jerryzh168 committed:
Update on "[quant][pt2e] Add reference representation for quantize_per_channel and dequantize_per_channel"


Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_quantize_dequantize_per_channel

Although right now it is not really testing things since there is some problem with dynamo export

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add reference representation for quantize_per_channel and dequantize_per_channel"


Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_quantize_dequantize_per_channel

Although right now it is not really testing things since there is some problem with dynamo export

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add reference representation for quantize_per_channel and dequantize_per_channel"


Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_quantize_dequantize_per_channel

Although right now it is not really testing things since there is some problem with dynamo export

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add reference representation for quantize_per_channel and dequantize_per_channel"


Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_quantize_dequantize_per_channel

Although right now it is not really testing things since there is some problem with dynamo export

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add reference representation for quantize_per_channel and dequantize_per_channel"


Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_quantize_dequantize_per_channel

Although right now it is not really testing things since there is some problem with dynamo export

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add reference representation for quantize_per_channel and dequantize_per_channel"


Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_quantize_dequantize_per_channel

Although right now it is not really testing things since there is some problem with dynamo export

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel reviewed (approved):
unblock

jerryzh168 commented:
@pytorchbot merge
